### PR TITLE
Make the LMDB backing store thread safe by using `RAWLock`s

### LINKED ISSUE #4408: Make the LMDB backing store thread safe
We should investigate whether the LMDB backing store is thread safe, and make it thread safe where it is not. One part of the LMDB backing store that is not thread safe is the `bsClose` function:

```haskell
    bsClose :: m ()
    bsClose = do
      guardDbClosed dbClosed
      Trace.traceWith dbTracer $ TDBClosing dbFilePath
      openHandles <- IOLike.readTVarIO dbOpenHandles
      for_ openHandles HD.bsvhClose
      IOLike.atomically $ IOLike.modifyTVar dbClosed (const True)
      liftIO $ LMDB.closeEnvironment dbEnv
      Trace.traceWith dbTracer $ TDBClosed dbFilePath
```

In this definition, the list of open handles is read atomically, but is only until after we have closed each handle that we declare the whole of the backing store as closed. This means that new value handles can be created that will not be closed automatically, even though the backing store will become closed at some point. Value handles that are still open after the backing store is closed will throw errors if we try to submit reads to them, since the underlying LMDB environment/database will have been closed.

### PR DESCRIPTION
**Pull request by jorisdral**
# Description

Resolves input-output-hk/ouroboros-consensus#52.

We make the LMDB backing store thread safe by using `RAWLock`s for the backing store and each backing store value handle. The locks keep track of the `Status` of the resource, which can either be `Open` or `Closed`. Once such a `Status` becomes `Closed`, all subsequent operations that try to acquire a lock will fail. See `Ouroboros.Consensus.Storage.LedgerDB.BackingStore.LMDB.Status`.

# Checklist

- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Commits have useful messages
    - [x] The documentation has been properly updated
    - [x] New tests are added if needed and existing tests are updated
    - [ ] Any changes affecting Consensus packages must have an entry in the appropriate `changelog.d` directory created using [`scriv`](https://github.com/input-output-hk/scriv). If in doubt, see the [Consensus release process](../ouroboros-consensus/docs/ReleaseProcess.md).
    - [ ] Any changes in the Consensus API (every exposed function, type or module) that has changed its name, has been deleted, has been moved, or altered in some other significant way must leave behind a `DEPRECATED` warning that notifies downstream consumers. If deprecating a whole module, remember to add it to `./scripts/ci/check-stylish.sh` as otherwise `stylish-haskell` would un-deprecate it.
    - [x] If this branch changes Network and has any consequences for downstream repositories or end users, said changes must be documented in [`interface-CHANGELOG.md`](../docs/interface-CHANGELOG.md)
    - [x] If serialization changes, user-facing consequences (e.g. replay from genesis) are confirmed to be intentional.
- Pull Request
    - [x] Self-reviewed the diff
    - [x] Useful pull request description at least containing the following information:
      - What does this PR change?
      - Why these changes were needed?
      - How does this affect downstream repositories and/or end-users?
      - Which ticket does this PR close (if any)? If it does, is it [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)?
    - [x] Reviewer requested

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs [hunk 1]
```diff
@@ -521,50 +516,57 @@ mkLMDBBackingStoreValueHandle db = do
...
 
+  vhStatus <- RAW.new Open
+
   let
-    bsvhClose :: m ()
-    bsvhClose = do
-      Trace.traceWith tracer TVHClosing
-      guardDbClosed dbClosed
-      guardBsvhClosed vhId dbOpenHandles
+    cleanup :: m ()
```
Comment (jorisdral) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs:522]: Might need some documentation

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB/Status.hs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+-- known to be 'Open', or an exception would have been thrown.
+withWriteAccess ::
+     (IOLike m, Exception e)
+  => RAWLock m Status
+  -> e                -- ^ The exception to throw
+  -> m (Status, a)    -- ^ Action to perform, possibly updating the 'Status'
+  -> m a
+withWriteAccess rawLock exc k =
+  RAW.withWriteAccess rawLock $ \case
+    Open  -> k
```
Comment (jorisdral) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB/Status.hs:52]: Alignment

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs [hunk 3]
```diff
@@ -88,8 +90,13 @@ data Db m l = Db {
   , dbBackingTables :: !(LedgerTables l LMDBMK)
   , dbFilePath      :: !FilePath
   , dbTracer        :: !(Trace.Tracer m TraceLMDB)
-  , dbClosed        :: !(IOLike.TVar m Bool)
-  , dbOpenHandles   :: !(IOLike.TVar m (Map Int (LMDBValueHandle l m)))
+    -- | Status of the LMDB backing store. When 'Closed', all backing store
+    -- (value handle) operations should fail.
```
Comment (dnadales) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs:94]: Should or will?

**Review by dnadales — COMMENTED** [review]
I wonder if we could:

- Use a simple `MVar` as lock (and spare us a new module)
- Revert the type of `dbOpenHandles`.

We should also revisit the deadlock scenario. The two locks are acquired in the same order everywhere, so I need to understand when a deadlock could occur :thinking:

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs [hunk 4]
```diff
@@ -88,8 +90,13 @@ data Db m l = Db {
...
+  , dbStatus        :: !(RAWLock m Status)
+    -- | Map of open value handles to cleanup functions. When closing the
+    -- backing store, these cleanup functions are used to ensure all value
+    -- handles are closed.
+  , dbOpenHandles   :: !(IOLike.TVar m (Map Int (m ())))
```
[the remark this answers is not in the document]
Comment (jorisdral) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs:99]: We get nested locks: `bsClose` acquires a write lock, then calls `bsvhClose` on a value handle, but `bsvhClose` acquires a read lock, and a deadlock is created

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs (hunk 3 above)
Comment (jorisdral) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs:94]: Will!

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs [hunk 5]
```diff
@@ -453,27 +450,25 @@ newLMDBBackingStoreInitialiser dbTracer limits sfs initFrom = do
    mkBackingStore :: Db m l -> LMDBBackingStore l m
    mkBackingStore db =
        let bsClose :: m ()
-           bsClose = do
-             guardDbClosed dbClosed
+           bsClose = withWriteAccess dbStatus DbErrClosed $ do
              Trace.traceWith dbTracer $ TDBClosing dbFilePath
              openHandles <- IOLike.readTVarIO dbOpenHandles
-             for_ openHandles HD.bsvhClose
```
[the remark this answers is not in the document]
Comment (jorisdral) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs:460]: You're right, I can change that

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jorisdral) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs:99]: See https://github.com/input-output-hk/ouroboros-network/pull/4498#discussion_r1163791583

**Comment by jorisdral** [comment]
> I wonder if we could:
> 
>     * Use a simple `MVar` as lock (and spare us a new module)

What I thought makes the `RAWLock` a good fit is that we're trying to keep the in-memory database status in sync with the status of the on-disk LMDB environment. See this haddock comment on the `RAWLock` type:

```haskell
-- It is important to realise that a RAW lock is intended to control access to
-- a piece of in-memory state that should remain in sync with some other state
-- that can only be modified using side-effects, e.g., the file system. If,
-- for example, you're only maintaining a counter shared by threads, then
-- simply use a 'TVar' or an 'MVar'.
```

In our case we do not only want to lock when we're closing, we also want to change some in-memory state from `Open` to `Closed`, such that all subsequent operations that try to acquire the lock will thrown an error. I think that could also easily be achieved with an `MVar` instead of the `TVar`-based `RAWLock` (and the `MVar` one might even be faster), but we would probably need auxiliary functions to get the "read-write lock with in-sync in-memory state" behaviour. Maybe I could create a separate module for an `MVar` based `RWLock`?

>     * Revert the type of `dbOpenHandles`.
> 
> We should also revisit the deadlock scenario. The two locks are acquired in the same order everywhere, so I need to understand when a deadlock could occur :thinking:

The deadlock does not involve both the `dbStatus` and `vhStatus` locks, but only the `dbStatus` lock. See https://github.com/input-output-hk/ouroboros-network/pull/4498#discussion_r1163791583

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs [hunk 6]
```diff
@@ -404,7 +401,7 @@ newLMDBBackingStoreInitialiser dbTracer limits sfs initFrom = do
    createOrGetDB = do
 
      dbOpenHandles <- IOLike.newTVarIO Map.empty
-     dbClosed      <- IOLike.newTVarIO False
+     dbStatus      <- RAW.new Open
```
[the remark this answers is not in the document]
Comment (jorisdral) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs:404]: See https://github.com/input-output-hk/ouroboros-network/pull/4498#issuecomment-1504898874

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs [hunk 7]
```diff
@@ -453,27 +450,25 @@ newLMDBBackingStoreInitialiser dbTracer limits sfs initFrom = do
    mkBackingStore :: Db m l -> LMDBBackingStore l m
    mkBackingStore db =
        let bsClose :: m ()
-           bsClose = do
-             guardDbClosed dbClosed
+           bsClose = withWriteAccess dbStatus DbErrClosed $ do
```
[the remark this answers is not in the document]
Comment (jorisdral) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs:453]: See https://github.com/input-output-hk/ouroboros-network/pull/4498#issuecomment-1504898874

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jorisdral) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs:99]: We should mention in the comment for `dbOpenHandles` why we don't just call `bsvhClose` on the value handles

[the remark this answers is not in the document]
Comment (dnadales) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs:99]: Thanks. Makes sense now. Make sure to mention this scenario in the comments of `dbOpenHandles`.

[the remark this answers is not in the document]
Comment (jorisdral) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs:99]: Done

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs (hunk 1 above)
Comment (jorisdral) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs:522]: Done

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs (hunk 3 above)
Comment (jorisdral) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs:94]: Changed

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs [hunk 8]
```diff
@@ -88,8 +90,13 @@ data Db m l = Db {
   , dbBackingTables :: !(LedgerTables l LMDBMK)
   , dbFilePath      :: !FilePath
   , dbTracer        :: !(Trace.Tracer m TraceLMDB)
-  , dbClosed        :: !(IOLike.TVar m Bool)
-  , dbOpenHandles   :: !(IOLike.TVar m (Map Int (LMDBValueHandle l m)))
+    -- | Status of the LMDB backing store. When 'Closed', all backing store
+    -- (value handle) operations should fail.
+  , dbStatus        :: !(RAWLock m Status)
```
[the remark this answers is not in the document]
Comment (jorisdral) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB.hs:95]: Done

**Comment by jorisdral** [comment]
Closed in favour of https://github.com/input-output-hk/ouroboros-consensus/pull/53, since we recently ported consensus-related code from `input-output-hk/ouroboros-network` to `input-output-hk/ouroboros-consensus` and all branches/PRs should be ported.

**Pull request closed without merging**